### PR TITLE
Tooltip: no-op when nested inside another Tooltip component

### LINKED ISSUE #56522: `Button` creates new HTML elements every time it internally flips between showing tooltips or not
_As discussed originally in https://github.com/WordPress/gutenberg/discussions/44735_

### The problem

The `Button` component can, [based on which props are passed](https://github.com/WordPress/gutenberg/blob/aa7cc4d0404e0a830144c57e5211335b48f9ba2d/packages/components/src/button/index.tsx#L195-L206), show a tooltip when hovered or focused.

This feature is implemented by[ wrapping conditionally the button's contents in a `Tooltip` component](https://github.com/WordPress/gutenberg/blob/aa7cc4d0404e0a830144c57e5211335b48f9ba2d/packages/components/src/button/index.tsx#L259-L292).

The current implementation causes React not to be able to [reconciliate](https://react.dev/learn/preserving-and-resetting-state) the Button's contents if the need for showing a tootlip changes at runtime. In return, this causes the current `button` (or `a`) element rendered in the DOM to be removed and replaced by a new element, which causes focus loss.

### Known workarounds and potential solutions

The most clear way to avoid the re-render is to make sure that the `Tooltip` component is always rendered (instead of being rendered conditionally). This can be achieved in a few ways:

1. by specifying the `showTooltip` prop on `Button` (although this will force the tooltip to be always visible) ([example](https://github.com/WordPress/gutenberg/blob/ca2ee4d6fa8c7044cd1d3045c0985a77659f44fa/packages/editor/src/components/post-saved-state/index.js#L155-L161))
2. by avoiding to trigger the internal tooltip in `Button`, and instead wrapping `Button` externally with a `Tooltip`  (tested in #56502)
3. by tweaking the `Button` component to internally always render a `Tooltip` component (proposed in #56490)

Approaches 1 and 2 can work well to fix specific instances of `Button`, while approach 3 would solve the problem fundamentally for all `Button` usages (although it has the potential to introduce a breaking change).

A 4th approach (yet to be tested) would be to try to use the `key` special prop as a way to flag to React that a given component is stable (and therefore doesn't need to be destroyed and re-created across re-renders).

### Plan of action

- [x] Merge #56502
- [x] Work on a workaround for all `Button` components
    - [x] Try out the `key` method #56527
    - [x] Alternatively, understand if the solution proposed in #56490 is viable
- [x] Potentially undo previously added workarounds (like in #56502)

### PR DESCRIPTION
**Pull request by ciampo**
Requires #57345
Pre-requisite for #56490

## What?

Allow nesting 2 or more `Tooltip` components without breaking the tooltip functionality. The contents of the tooltip will be the ones defined in the outer-most `Tooltip` instance.

## Why?

In #56490, we're trying to fix [an issue](https://github.com/WordPress/gutenberg/issues/56522) where `Button` could trash and re-create its children when the internal tooltip logic flips from showing to hiding the tootlip (or vice-versa).

The solution proposed in #56490 involves always rendering (in the react sense) the `Tooltip` component inside `Button`. But doing so would cause breakages, since the `Tooltip` component doesn't support that scenario.

This PR allows that scenario to work without causing breakages.

## How?

By using the internal context system to inform instances of `Tooltip` whether they are being rendered inside another instance of `Tooltip`. When that happens, the `Tooltip` component only renders the anchor and ignores the tooltip part.

## Testing Instructions

- Test the temporary "Nested" storybook example
- Make sure unit tests pass

### CODE REVIEW

**Comment by ciampo** [comment]
Before moving any further, I'd like to get a quick check from @diegohaz on the chosen approach. Diego, here is a quick recap:
- it was initially noticed that when `Button`'s internal logic flips between showing/not showing a tooltip, React is forced to trash/re-create the underlying HTML element, causing a loss of focus (see https://github.com/WordPress/gutenberg/issues/56522)
- After trying a few approaches, the only one that seems to work reliably is to always render a `Tooltip` component inside the `Button` element (#56490), although this approach caused issues if `Button` is already nested inside a `Tooltip` components (see https://github.com/WordPress/gutenberg/pull/56490#issuecomment-1825498243).
- Therefore, we decided to look into tweaking `Tooltip` so that if multiple instances are nested into each other, only the outer instance actually renders a tooltip (this PR)

Please let me know if this approach makes sense to you, or if there are better ways to handle this in `ariakit`. Thank you 🙏

**Comment by diegohaz** [comment]
I believe that's reasonable. If I'm interpreting the solution correctly, the only issue I foresee is if users typically display portaled content within those buttons. For example:

```jsx
<Tooltip>
  <Button />
  <Modal>
    <Tooltip>
      <Button />
    </Tooltip>
  </Modal>
</Tooltip>
```

If this scenario is unlikely, then I think it's perfectly fine.

File: packages/components/src/tooltip/index.tsx
Comment (ciampo) [inline packages/components/src/tooltip/index.tsx]: Adding the early return when `isNestedInParentTooltip` causes ESlint to complain that `tooltipStore` may be unused before `return`. But hooks can't be run conditionally, and therefore I disabled the ESlint warning.

File: packages/components/src/tooltip/index.tsx
Comment (ciampo) [inline packages/components/src/tooltip/index.tsx:40]: For the nesting to work correctly, `Tooltip` needs to forward the ref down to its anchor.

File: packages/components/src/tooltip/stories/index.story.tsx
Comment (ciampo) [inline packages/components/src/tooltip/stories/index.story.tsx]: This example is only around for ease of reviewing / testing. I will remove it before merging.

**Comment by ciampo** [comment]
Rebased and marked as ready for review after #57345 got merged.

File: packages/components/src/tooltip/index.tsx
Comment (tyxla) [inline packages/components/src/tooltip/index.tsx]: This ESLint error can be circumvented by making it clearer that we're using a hook.

Something like this will do the job:

```suggestion
	const useAriakitTooltipStore = Ariakit.useTooltipStore;
	const tooltipStore = useAriakitTooltipStore( {
```

Feel free to also add a comment that explains why we're doing this.

File: packages/components/src/tooltip/stories/index.story.tsx
Comment (tyxla) [reply packages/components/src/tooltip/stories/index.story.tsx]: Should we keep a similar example in Storybook to demonstrate the `isNestedInParentTooltip` prop though?

File: packages/components/src/tooltip/test/index.tsx
Comment (tyxla) [inline packages/components/src/tooltip/test/index.tsx]: Did you mean `toBeVisible()`?

```suggestion
			).toBeVisible();
```

File: packages/components/src/tooltip/index.tsx
Comment (tyxla) [inline packages/components/src/tooltip/index.tsx]: Naming nit: the "parent" term seems redundant, I'd go with something like `isNestedTooltip`. To me it's clear that if it's a nested tooltip, then it's in a parent tooltip. WDYT?

File: packages/components/src/tooltip/index.tsx
Comment (tyxla) [reply packages/components/src/tooltip/index.tsx]: See my comment a few lines below on how to improve this.

File: packages/components/src/tooltip/test/index.tsx
Comment (tyxla) [inline packages/components/src/tooltip/test/index.tsx]: Nit: I think we should use "nested" instead of "inner" for better clarity. 

```suggestion
		it( 'should render the outer tooltip and ignore nested tooltips', async () => {
```

File: packages/editor/src/components/post-saved-state/test/__snapshots__/index.js.snap
Comment (tyxla) [inline packages/editor/src/components/post-saved-state/test/__snapshots__/index.js.snap]: This `undefined` classname seems off.  Is it signaling a poor use of the `className` prop inside `<PostSavedState />`?

File: packages/components/CHANGELOG.md
Comment (tyxla) [inline packages/components/CHANGELOG.md:35]: Is this actually a breaking change? Could anyone have relied on previously allowed nested tooltips before?

File: packages/components/src/tooltip/index.tsx
Comment (tyxla) [inline packages/components/src/tooltip/index.tsx]: I also think we could use the opportunity to document and explain the rationale behind how we handle nested tooltip behavior in the component README.

File: packages/components/src/tooltip/index.tsx
Comment (tyxla) [inline packages/components/src/tooltip/index.tsx]: Any potential core use cases with nested behavior that could be affected by this change?

Comment (tyxla) [inline packages/components/src/tooltip/index.tsx]: Did we also mean to trigger a console warning in that scenario (as per your [earlier suggestion](https://github.com/WordPress/gutenberg/pull/56490#issuecomment-1826427624))?

File: packages/components/src/tooltip/index.tsx
Comment (ciampo) [reply packages/components/src/tooltip/index.tsx]: [`6e852d5` (#57202)](https://github.com/WordPress/gutenberg/pull/57202/commits/6e852d53c3e29de89422c072df2beab27017b925)

File: packages/components/src/tooltip/stories/index.story.tsx
Comment (ciampo) [reply packages/components/src/tooltip/stories/index.story.tsx]: [`1b7f1a1` (#57202)](https://github.com/WordPress/gutenberg/pull/57202/commits/1b7f1a12606506bf48878ef3b2bbe9a930a313c1)

File: packages/components/src/tooltip/index.tsx
Comment (ciampo) [reply packages/components/src/tooltip/index.tsx]: [`08cfb81` (#57202)](https://github.com/WordPress/gutenberg/pull/57202/commits/08cfb8195dc757b7ebb0dd288518e1e1fffe09a7)

File: packages/components/src/tooltip/test/index.tsx
Comment (ciampo) [reply packages/components/src/tooltip/test/index.tsx]: [`697b30b` (#57202)](https://github.com/WordPress/gutenberg/pull/57202/commits/697b30b3a841b4f151f2e5a5c3c97b2cf8c25b3a)

File: packages/components/src/tooltip/test/index.tsx
Comment (ciampo) [reply packages/components/src/tooltip/test/index.tsx]: [`edb03e7` (#57202)](https://github.com/WordPress/gutenberg/pull/57202/commits/edb03e77cf32aa494870bd8bfb61c2d8adc41e8e)

File: packages/editor/src/components/post-saved-state/test/__snapshots__/index.js.snap
Comment (ciampo) [reply packages/editor/src/components/post-saved-state/test/__snapshots__/index.js.snap]: Well spotted — that was a consequence of using the `useContext` hook, which actually made me realize that we should probably forward the `restProps` to the tooltip, and not the anchor. Worked on it in [`d52af82` (#57202)](https://github.com/WordPress/gutenberg/pull/57202/commits/d52af821c9e4b7db237c997b7b8a70eba10709c1), which also allows us to revert all of the snapshot changes

File: packages/components/src/tooltip/index.tsx
Comment (ciampo) [reply packages/components/src/tooltip/index.tsx]: Added a section in the README [`7fbfcbe` (#57202)](https://github.com/WordPress/gutenberg/pull/57202/commits/7fbfcbef86c32765dbbdfd8bf27ccc60186894b9)

File: packages/components/src/tooltip/index.tsx
Comment (ciampo) [reply packages/components/src/tooltip/index.tsx]: I thought about it, and then I decided _not_ to add a warning as it has the potential to pollute the browser's console‚ especially if we're going to implement the solution proposed in #56490, which purposefully makes use of this nested tooltip logic.

Comment (ciampo) [reply packages/components/src/tooltip/index.tsx]: I can't think of any, if anything because, prior to this PR, I don't think that nesting `Tooltip` components would not work correctly.

File: packages/components/CHANGELOG.md
Comment (ciampo) [reply packages/components/CHANGELOG.md:35]: I don't think so, as I don't think that nesting `Tooltip`s would have worked before.

File: packages/components/src/tooltip/stories/index.story.tsx
Comment (tyxla) [inline packages/components/src/tooltip/stories/index.story.tsx]: This comment seems unfinished.

**Review by tyxla — APPROVED** [review]
Nothing further from me on this one. Great work 🚀

File: packages/components/src/tooltip/stories/index.story.tsx
Comment (ciampo) [reply packages/components/src/tooltip/stories/index.story.tsx]: Well spotted, I updated it with the same text that I added to the README.

**Comment by ciampo** [comment]
Rebased to include the latest updated to `ariakit` version, will merge once CI checks pass 🟢

File: packages/components/src/tooltip/index.tsx
Comment (tyxla) [reply packages/components/src/tooltip/index.tsx]: We're reverting this now that it's no longer conditionally called: https://github.com/WordPress/gutenberg/pull/61858

**Pull request merged**
